internal/controller/sparkconnect/reconciler.go

```diff
@@ -0,0 +1,565 @@
+			}
+			container.Image = *conn.Spec.Image
+		}
+
+		// Setup entrypoint.
+		container.Command = []string{"bash", "-c"}
+		args, err := buildStartConnectServerArgs(conn)
+		if err != nil {
+			return fmt.Errorf("failed to build spark connection args: %v", err)
+		}
+		container.Args = []string{strings.Join(args, " ")}
+
+		// Setup environment variables.
+		container.Env = append(
+			container.Env,
+			corev1.EnvVar{
+				Name: "POD_IP",
+				ValueFrom: &corev1.EnvVarSource{
+					FieldRef: &corev1.ObjectFieldSelector{
+						FieldPath: "status.podIP",
+					},
+				},
+			},
+			corev1.EnvVar{
+				Name:  common.EnvSparkNoDaemonize,
+				Value: "true",
+			},
+		)
+
+		// Setup volumes and volumeMounts.
+		container.VolumeMounts = append(
+			container.VolumeMounts,
+			corev1.VolumeMount{
+				Name:      "spark-conf",
+				SubPath:   ExecutorPodTemplateFileName,
+				MountPath: fmt.Sprintf("/tmp/spark/%s", ExecutorPodTemplateFileName),
+				ReadOnly:  true,
+			},
+		)
+
+		pod.Spec.Volumes = append(
+			pod.Spec.Volumes,
+			corev1.Volume{
+				Name: "spark-conf",
+				VolumeSource: corev1.VolumeSource{
+					ConfigMap: &corev1.ConfigMapVolumeSource{
+						LocalObjectReference: corev1.LocalObjectReference{
+							Name: GetConfigMapName(conn),
+						},
+						Items: []corev1.KeyToPath{
+							{
+								Key:  ExecutorPodTemplateFileName,
+								Path: ExecutorPodTemplateFileName,
+							},
+						},
+					},
+				},
+			},
+		)
+	}
+
+	// Set controller owner reference on server pod.
+	if err := ctrl.SetControllerReference(conn, pod, r.scheme); err != nil {
+		return fmt.Errorf("failed to set controller reference: %v", err)
+	}
+
+	if pod.Labels == nil {
+		pod.Labels = map[string]string{}
+	}
+	pod.Labels[common.LabelLaunchedBySparkOperator] = "true"
+	pod.Labels[common.LabelSparkConnName] = conn.Name
+
+	return nil
+}
+
+// createOrUpdateServerService creates or updates the server service for the SparkConnect resource.
+func (r *Reconciler) createOrUpdateServerService(ctx context.Context, conn *v1alpha1.SparkConnect) error {
+	logger := ctrl.LoggerFrom(ctx)
+	logger.V(1).Info("Create or update server service")
+
+	svc := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      GetServerServiceName(conn),
+			Namespace: conn.Namespace,
+		},
+	}
+
+	_, err := controllerutil.CreateOrUpdate(ctx, r.client, svc, func() error {
+		if err := r.mutateServerService(ctx, conn, svc); err != nil {
+			return fmt.Errorf("failed to mutate server service: %v", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create or update server service: %v", err)
+	}
+
+	// Update SparkConnect status.
+	conn.Status.Server.ServiceName = svc.Name
+
+	return nil
+}
+
+// mutateServerService mutates the server service for the SparkConnect resource.
+func (r *Reconciler) mutateServerService(ctx context.Context, conn *v1alpha1.SparkConnect, svc *corev1.Service) error {
+	if svc.CreationTimestamp.IsZero() {
```

Comment:
Do we want to ensure this is applied every reconciliation loop? Not just the first time it's created.